# OilyGiant
## Análisis de pozos petroleros

La compañía de extracción de petróleo OilyGiant está buscando los mejores lugares donde abrir 200 pozos nuevos de petróleo. Con un modelo de Regresión Lineal se pretende encontrar la mejor región cuyos 200 pozos más productivos generen la mayor ganancia para la empresa y tenga el menor riesgo de pérdida. 

## 1. Preprocesamiento de datos: 

In [1]:
#Librerías utilizadas.

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

In [2]:
# Uso en el ordenador.

#data_0 = pd.read_csv('./geo_data_0.csv')

#data_1 = pd.read_csv('./geo_data_1.csv')

#data_2 = pd.read_csv('./geo_data_2.csv')

# Revisión externa.

data_0 = pd.read_csv('/datasets/geo_data_0.csv')

data_1 = pd.read_csv('/datasets/geo_data_1.csv')

data_2 = pd.read_csv('/datasets/geo_data_2.csv')

In [3]:
data_0.head()

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

In [4]:
data_1.head()

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

In [5]:
data_2.head()

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

In [6]:
data_0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [7]:
data_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [8]:
data_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [9]:
data_0.describe()

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.500419       0.250143       2.502647      92.500000
std         0.871832       0.504433       3.248248      44.288691
min        -1.408605      -0.848218     -12.088328       0.000000
25%        -0.072580      -0.200881       0.287748      56.497507
50%         0.502360       0.250252       2.515969      91.849972
75%         1.073581       0.700646       4.715088     128.564089
max         2.362331       1.343769      16.003790     185.364347

In [10]:
data_1.describe()

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        1.141296      -4.796579       2.494541      68.825000
std         8.965932       5.119872       1.703572      45.944423
min       -31.609576     -26.358598      -0.018144       0.000000
25%        -6.298551      -8.267985       1.000021      26.953261
50%         1.153055      -4.813172       2.011479      57.085625
75%         8.621015      -1.332816       3.999904     107.813044
max        29.421755      18.734063       5.019721     137.945408

In [11]:
data_2.describe()

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.002023      -0.002081       2.495128      95.000000
std         1.732045       1.730417       3.473445      44.749921
min        -8.760004      -7.084020     -11.970335       0.000000
25%        -1.162288      -1.174820       0.130359      59.450441
50%         0.009424      -0.009482       2.484236      94.925613
75%         1.158535       1.163678       4.858794     130.595027
max         7.238262       7.844801      16.739402     190.029838

In [12]:
# Búsqueda de información duplicada. 

print(data_0.duplicated().sum())
print()
print(data_1.duplicated().sum())
print()
print(data_2.duplicated().sum())

0

0

0


In [13]:
# Evidencia del tamaño de los datos para separlos correctamente durante la elaboración del modelo de ML. 

print(data_0.shape)
print()
print(data_1.shape)
print()
print(data_2.shape)

(100000, 5)

(100000, 5)

(100000, 5)


In [14]:
#Eliminación de la columna que no aporta al modelo de machine learning.

data_0 = data_0.drop(['id'], axis=1)

data_1 = data_1.drop(['id'], axis=1)

data_2 = data_2.drop(['id'], axis=1)

Tras un sondeo de las bases de datos se puede observar que todos los datos son del tipo correcto, no hay datos ausentes ni duplicados. Además, contamos con 3 bases con la misma cantidad de datos y distribución, por todo lo anterior es que podemos proceder a hacer la división de los datos y entrenar los modelos. 

## 2. División de los datos y entrenamiento de modelos

### 2.1 División de bases de datos:

In [15]:
#División de las tres bases de datos.

data_train_0, data_valid_0 = train_test_split(data_0, test_size = 0.25, train_size = 0.75, random_state = 123)

data_train_1, data_valid_1 = train_test_split(data_1, test_size = 0.25, train_size = 0.75, random_state = 123)

data_train_2, data_valid_2 = train_test_split(data_2, test_size = 0.25, train_size = 0.75, random_state = 123)

In [16]:
# Revisión de la división de datos.

print(f'Region 0 dt:', data_train_0.shape, data_valid_0.shape)
print()
print(f'Region 1 dt:',data_train_0.shape, data_valid_0.shape)
print()
print(f'Region 2 dt:',data_train_0.shape, data_valid_0.shape)

Region 0 dt: (75000, 4) (25000, 4)

Region 1 dt: (75000, 4) (25000, 4)

Region 2 dt: (75000, 4) (25000, 4)


### 2.2 Entrenamiento del modelo de Regresión Lineal para todas las regiones: 

In [17]:
# Modelo región 0.

train_features_0 = data_train_0.drop(['product'], axis=1)
train_target_0 = data_train_0['product']

valid_features_0 = data_valid_0.drop(['product'], axis=1)
valid_target_0 = data_valid_0['product']

model = LinearRegression()

model.fit(train_features_0, train_target_0)

predictions_valid_0 = model.predict(valid_features_0)

mse_0 = mean_squared_error(valid_target_0, predictions_valid_0)

print(f"""
      Average volume predicted:, {predictions_valid_0.mean().round(2)}
      MSE region 0:, {mse_0}
      RMSE region 0: {mse_0 ** 0.5}
      """)



      Average volume predicted:, 92.55
      MSE region 0:, 1417.3615751967836
      RMSE region 0: 37.64786282376177
      


In [18]:
# Modelo región 1.

train_features_1 = data_train_1.drop(['product'], axis=1)
train_target_1 = data_train_1['product']

valid_features_1 = data_valid_1.drop(['product'], axis=1)
valid_target_1 = data_valid_1['product']

model = LinearRegression()

model.fit(train_features_1, train_target_1)

predictions_valid_1 = model.predict(valid_features_1)

mse_1 = mean_squared_error(valid_target_1, predictions_valid_1)

print(f"""
      Average volume predicted:, {predictions_valid_1.mean().round(2)}
      MSE region 1:, {mse_1}
      RMSE region 1: {mse_1 ** 0.5}
      """)



      Average volume predicted:, 69.28
      MSE region 1:, 0.8017661964648853
      RMSE region 1: 0.8954139804944333
      


In [19]:
# Modelo región 2.

train_features_2 = data_train_2.drop(['product'], axis=1)
train_target_2 = data_train_2['product']

valid_features_2 = data_valid_2.drop(['product'], axis=1)
valid_target_2 = data_valid_2['product']

model = LinearRegression()

model.fit(train_features_2, train_target_2)

predictions_valid_2 = model.predict(valid_features_2)

mse_2 = mean_squared_error(valid_target_2, predictions_valid_2)

print(f"""
      Average volume predicted:, {predictions_valid_2.mean().round(2)}
      MSE region 2:, {mse_2}
      RMSE region 2: {mse_2 ** 0.5}
      """)



      Average volume predicted:, 95.1
      MSE region 2:, 1610.2587969766075
      RMSE region 2: 40.12803006598514
      


### 2.3 Prueba de cordura para los modelos:

In [20]:
# Función para aplicar la prueba de cordura en todas las regiones.

def sanity_test(train_data, valid_data):

    train_features = train_data.drop(['product'], axis=1)
    train_target = train_data['product']

    valid_features = valid_data.drop(['product'], axis=1)
    valid_target = valid_data['product']

    model = LinearRegression()

    model.fit(train_features, train_target)
    
    mean_predictions_valid = pd.Series(train_target.mean(), index = valid_target.index)

    mse_mean = mean_squared_error(valid_target, mean_predictions_valid)

    print('Sanity test results')
    print('Average MSE =', mse_mean)
    print('Average RMSE =', mse_mean ** 0.5)
    
    return sanity_test

### 2.4 Comparativa y muestra de resultados de los modelos:

In [21]:
print('Region 0')
print(f"""
Average volume predicted:, {predictions_valid_0.mean().round(2)}
MSE region 0:, {mse_0}
RMSE region 0: {mse_0 ** 0.5}
      """)
sanity_test_0 = sanity_test(data_train_0, data_valid_0)
print()

print('Region 1')
print(f"""
Average volume predicted:, {predictions_valid_1.mean().round(2)}
MSE region 1:, {mse_1}
RMSE region 1: {mse_1 ** 0.5}
      """)
sanity_test_1 = sanity_test(data_train_1, data_valid_1)
print()

print('Region 2')
print(f"""
Average volume predicted:, {predictions_valid_2.mean().round(2)}
MSE region 2:, {mse_2}
RMSE region 2: {mse_2 ** 0.5}
      """)
sanity_test_2 = sanity_test(data_train_2, data_valid_2)

Region 0

Average volume predicted:, 92.55
MSE region 0:, 1417.3615751967836
RMSE region 0: 37.64786282376177
      
Sanity test results
Average MSE = 1972.3302703693828
Average RMSE = 44.41092512399829

Region 1

Average volume predicted:, 69.28
MSE region 1:, 0.8017661964648853
RMSE region 1: 0.8954139804944333
      
Sanity test results
Average MSE = 2099.7296790793853
Average RMSE = 45.82280741158692

Region 2

Average volume predicted:, 95.1
MSE region 2:, 1610.2587969766075
RMSE region 2: 40.12803006598514
      
Sanity test results
Average MSE = 1995.7302728988466
Average RMSE = 44.67359704455023


### 2.5 Conclusiones parciales:

Con los anteriores resultados podemos observar que el modelo de regresión lineal en cada región es más efectivo que los resultados de la prueba de cordura donde se trabaja con los datos promediados. 

Además, podemos extraer otros datos como las predicciones frente a la producción de miles de barriles en cada región, en este caso, se estima que la mayor producción se genera en la región 2.

## 3. Primeros pasos para el cálculo de ganancias

### 3.1 Preparación de los datos:

In [22]:
# Valores necesarios para los cálculos: 

investment = 100000000

oil_wells = 200

oil_wells_per_region = 10000

min_product = 111.1

income_per_barrel = 4.5

income_per_unit = 4500

investment_per_well = investment / oil_wells

min_product_per_region = oil_wells_per_region * min_product

min_income_per_region = min_product_per_region * income_per_unit

print(f'Investment per well:' , investment_per_well, 'dollars')
print()
print(f'Min production of barrels per well:' , min_product_per_region, 'of barrels')
print()
print(f'Minimum income from a profitable well:' , min_income_per_region, 'dollars')

Investment per well: 500000.0 dollars

Min production of barrels per well: 1111000.0 of barrels

Minimum income from a profitable well: 4999500000.0 dollars


### 3.2 Cálculos por región:

In [23]:
# Región 0:

av_production_region_0 = data_0['product'].mean()

av_income_region_0 = av_production_region_0 * income_per_unit


print(f'Average production of region 0 per well:', av_production_region_0.round(2), 'thousands of barrels')
print()
print(f'Average income of region 0 per well:', av_income_region_0.round(2), 'dollars')

Average production of region 0 per well: 92.5 thousands of barrels

Average income of region 0 per well: 416250.0 dollars


In [24]:
# Región 1:

av_production_region_1 = data_1['product'].mean()

av_income_region_1 = av_production_region_1 * income_per_unit


print(f'Average production of region 1 per well:', av_production_region_1.round(2), 'thousands of barrels')
print()
print(f'Average income of region 1 per well:', av_income_region_1.round(2), 'dollars')

Average production of region 1 per well: 68.83 thousands of barrels

Average income of region 1 per well: 309712.5 dollars


In [25]:
# Región 2:

av_production_region_2 = data_2['product'].mean()

av_income_region_2 = av_production_region_2 * income_per_unit


print(f'Average production of region 2 per well:', av_production_region_2.round(2), 'thousands of barrels')
print()
print(f'Average income of region 2 per well:', av_income_region_2.round(2), 'dollars')

Average production of region 2 per well: 95.0 thousands of barrels

Average income of region 2 per well: 427500.0 dollars


### 3.3 Comparativa de las regiones: 

In [26]:
# Producción promedio de miles de barriles por región

print(f'Average production of region 0 per well:', av_production_region_0.round(2), 'thousands of barrels')
print()
print(f'Average production of region 1 per well:', av_production_region_1.round(2), 'thousands of barrels')
print()
print(f'Average production of region 2 per well:', av_production_region_2.round(2), 'thousands of barrels')

Average production of region 0 per well: 92.5 thousands of barrels

Average production of region 1 per well: 68.83 thousands of barrels

Average production of region 2 per well: 95.0 thousands of barrels


### 3.4 Conclusiones parciales:

Tras calcular la producción promedio de cada región podemos ver que, cuando se trata del promedio, ninguna región supera los 111.1 miles de barriles mínimos que se deben extraer para que un pozo sea rentable; sin embargo, la producción de la región 0 y 2 son las que más se acercan. 

## 4. Calculando las ganancias de un conjunto de 200 pozos de petróleo por región

### 4.1 Elección de los mejores pozos por región: 

In [27]:
# Creación de listas con los datos del modelo de ML.

region_0_pred = pd.Series(predictions_valid_0, index = valid_target_0.index).sort_values(ascending = False)

region_0_pred.head()

93073    180.192010
78948    175.758271
95537    171.721784
74946    168.782147
95898    168.271148
dtype: float64

In [28]:
region_1_pred = pd.Series(predictions_valid_1, index = valid_target_0.index).sort_values(ascending = False)

region_1_pred.head()

55165    139.906443
42738    139.853002
19483    139.724252
14041    139.704776
33785    139.672587
dtype: float64

In [29]:
region_2_pred = pd.Series(predictions_valid_2, index = valid_target_0.index).sort_values(ascending = False)

region_2_pred.head()

6647     174.573336
98619    166.796364
29787    164.518817
64380    164.513090
17415    163.097615
dtype: float64

### 4.2 Calculando los beneficios por región:

In [30]:
# Función para calcular el ingreso por región tomando sus 200 pozos con mayor producción. 

def profit(predictions_valid, target, n_wells, print_results = True):
    region_pred = pd.DataFrame({'predicted_product': predictions_valid, 'target_product': target})
                               
    top_200 = region_pred.sort_values(by = 'predicted_product', 
                                                          ascending = False).head(n_wells)
    
    investment = 100000000

    total_reserved_volume = top_200['predicted_product'].sum()
    potential_profit = (total_reserved_volume * 4500) - investment
    
    if print_results:
        print(f"""
        Total product: {total_reserved_volume.round(2)} thousands of barrels
        Potencial profit: {potential_profit.round(2)} dollars
        """)
        
    return potential_profit


In [31]:
print('Region 0:')
profit(predictions_valid_0, valid_target_0, 200)
print()

print('Region 1:')
profit(predictions_valid_1, valid_target_1, 200)
print()

print('Region 2:')
profit(predictions_valid_2, valid_target_2, 200)

Region 0:

        Total product: 31034.79 thousands of barrels
        Potencial profit: 39656545.12 dollars
        

Region 1:

        Total product: 27755.76 thousands of barrels
        Potencial profit: 24900928.92 dollars
        

Region 2:

        Total product: 29943.44 thousands of barrels
        Potencial profit: 34745494.0 dollars
        


34745493.996762305

### 4.3 Conclusiones parciales: 

Gracias a los anteriores cálculos podemos observar que las predicciones frente a la producción de los 200 mejores pozos de cada región superan la cantidad mínima para hacerlos rentables. 

Según las predicciones: 

- La región con mayor producción y ganancias es la región 0, con una producción total de 31034.79 miles de barriles y un ingreso de 39,656,545.12 dólares con tan solo 200 pozos.

## 5. Riesgos y ganancias por región

In [32]:
def bootstrapping(target, predictions, sample_size):
    
    state = np.random.RandomState(123)
    
    profit_values = []

    for i in range(1000):
        target_subsample = target.sample(sample_size, replace = True, random_state = state)
        predict_subsample = predictions[target_subsample.index]
        
        profit_values.append(profit(target_subsample, predict_subsample, 200, print_results = False))
    
    profit_values = pd.Series(profit_values)
    
    av_profit = profit_values.mean()
    lower = profit_values.quantile(.025)
    upper = profit_values.quantile(.975)
    loss_risk = profit_values[profit_values < 0].count() / len(profit_values)
    risk_of_loss_percentage = loss_risk * 100
    
    print('Average Profit:', av_profit.round(2))
    print('Lower limit:', lower.round(2))
    print('Upper limit:', upper.round(2)),
    print('Loss risk:', loss_risk)
    print('Loss risk percentage:', risk_of_loss_percentage)
        
    return bootstrapping


In [33]:
print('Region 0')
bootstrapping(valid_target_0, region_0_pred, 500)
print()
print('Region 1')
bootstrapping(valid_target_1, region_1_pred, 500)
print()
print('Region 2')
bootstrapping(valid_target_2, region_2_pred, 500)
print()

Region 0
Average Profit: 24307011.26
Lower limit: 20320645.28
Upper limit: 28408608.29
Loss risk: 0.0
Loss risk percentage: 0.0

Region 1
Average Profit: 4802945.82
Lower limit: 588932.62
Upper limit: 8744248.19
Loss risk: 0.009
Loss risk percentage: 0.8999999999999999

Region 2
Average Profit: 26002682.74
Lower limit: 21950639.82
Upper limit: 29716725.68
Loss risk: 0.0
Loss risk percentage: 0.0



## 6. Conclusiones

Tras un muestreo con la técnica bootstrapping se puede observar que: 

1. La única región con riesgo de pérdida es la 1, sin embargo esta no es tan significativa. 
2. La región con el margen de ganancia más prometedor es la 2, además el promedio de ganancias es el más alto al ser 26,002,682.82 dólares.

Es por lo anterior que podemos concluir que la mejor región para abrir 200 pozos nuevos de petroleo es la 2, cambiando de parecer frente a los primeros cálculos que indicaban que los mayores ingresos se darían en la región 0, esta pasa entonces a ser la segunda mejor opción y se confirma de nuevo la idea de descartar la región 1 como opción. 